src/CLRunner.ts

```diff
@@ -335,6 +335,110 @@ export class CLRunner {
+
+        // Don't save logs on server if custom storage was provided
+        if (ConversationLearner.logStorage) {
+            createParams.saveToLog = false
+        }
+
+        const session = await this.clClient.StartSession(appId, createParams as CLM.SessionCreateParams)
+
+        // If using customer storage add to log storage
+        if (ConversationLearner.logStorage) {
+            // For self-hosted log storage logDialogId is sessionId
+            session.logDialogId = session.sessionId
+            const logDialog: CLM.LogDialog = {
+                logDialogId: session.logDialogId,
+                packageId: session.packageId,
+                rounds: [],
+                initialFilledEntities: [],
+                targetTrainDialogIds: [],
+                createdDateTime: new Date().toJSON(),
+                // Start out the same.  End is updated when dialog is edited
+                dialogBeginDatetime: new Date().toJSON(),
+                dialogEndDatetime: new Date().toJSON(),
+                lastModifiedDateTime: new Date().toJSON(),
+                metrics: ""
+            }
+            await ConversationLearner.logStorage.Add(appId, logDialog)
+        }
+        return session
+    }
+
+    private async SessionExtract(appId: string, sessionId: string, userInput: CLM.UserInput): Promise<CLM.ExtractResponse> {
+        const stepBeginDatetime = new Date().toJSON()
+        const extractResponse = await this.clClient.SessionExtract(appId, sessionId, userInput)
+        const stepEndDatetime = new Date().toJSON()
+
+        // Add to dev's log storage account (if it exists)
+        if (ConversationLearner.logStorage) {
+            // For local stroate logDialogId = sessionId
+            const logDialogId = sessionId
+
+            // Append an extractor step to already existing log dialog
+            const logDialog: CLM.LogDialog | undefined = await ConversationLearner.logStorage.Get(appId, logDialogId)
+            if (!logDialog) {
+                throw new Error(`Log Dialog does not exist App:${appId} Id:${logDialogId}`)
+            }
+            const newRound: CLM.LogRound = {
+                extractorStep: { ...extractResponse, stepBeginDatetime, stepEndDatetime },
+                scorerSteps: []
+            }
+            logDialog.rounds.push(newRound)
+            await ConversationLearner.logStorage.Replace(appId, logDialog)
+        }
+        return extractResponse
+    }
+
+    private async SessionScore(appId: string, sessionId: string, scoreInput: CLM.ScoreInput): Promise<CLM.ScoreResponse> {
+        const stepBeginDatetime = new Date().toJSON()
+        const scoreResponse = await this.clClient.SessionScore(appId, sessionId, scoreInput)
+
+        // Add to dev's log storage account (if it exists)
+        if (ConversationLearner.logStorage) {
+            // For self-hosted storage logDialogId is sessionId
```

Comment:
Suggest using standard nomenclature - "local storage" or "self-hosted storage", not both.